core/src/bytecode/ast/compat.rs

```diff
@@ -1252,3 +1366,178 @@ impl<'ast> FromAst<&'ast Ast<'ast>> for term::RichTerm {
         FromAst::from_ast(*ast)
     }
 }
+
+/// Convert a record definition to a mainline recursive record definition (with static and dynamic
+/// fields). If a field is defined several times, the different definitions are merged, statically
+/// if possible.
+
+impl<'ast> FromAst<Record<'ast>>
+    for (
+        term::record::RecordData,
+        Vec<(term::RichTerm, term::record::Field)>,
+    )
+{
+    fn from_ast(record: &Record<'ast>) -> Self {
```

Comment:
This is mostly `parser::utils::build_record` adapted to take a new AST as an input.